File: web/lib/devig.ts
// No-vig (devigged) fair-probability math for two-way totals.
// Exact mirror of beatvegas/devig.py — keep the two in sync (same test vectors).
// Devig outputs are MARKET signal: use only for CLV, line-check EV, and staking
// — never feed them into the market-blind BV regressor.

export function americanToDecimal(price: number): number {
  return price < 0 ? 1 + 100 / Math.abs(price) : 1 + price / 100;
}

export function americanToProb(price: number): number {
  return 1 / americanToDecimal(price);
}

function multiplicative(pOver: number, pUnder: number): [number, number] {
  const total = pOver + pUnder;
  return [pOver / total, pUnder / total];
}

// Find k such that pOver**k + pUnder**k == 1, via bisection.
function power(pOver: number, pUnder: number): [number, number] {
  let lo = 0;
  let hi = 10;
  for (let i = 0; i < 60; i++) {
    const k = (lo + hi) / 2;
    const s = pOver ** k + pUnder ** k;
    if (s > 1) lo = k;
    else hi = k;
  }
  const k = (lo + hi) / 2;
  return [pOver ** k, pUnder ** k];
}

// Shin two-outcome devig: solve for the insider proportion z in (0, 0.5) by
// bisection on sqrt(z^2+4(1-z)a) + sqrt(z^2+4(1-z)b) = 2.
function shin(pOver: number, pUnder: number): [number, number] {
  const booksum = pOver + pUnder;
  if (booksum <= 1) return multiplicative(pOver, pUnder);
  const a = (pOver * pOver) / booksum;
  const b = (pUnder * pUnder) / booksum;
  const s = (z: number) =>
    Math.sqrt(z * z + 4 * (1 - z) * a) + Math.sqrt(z * z + 4 * (1 - z) * b);
  let lo = 0;
  let hi = 0.5;
  for (let i = 0; i < 60; i++) {
    const z = (lo + hi) / 2;
    if (s(z) > 2) lo = z;
    else hi = z;
  }
  const z = (lo + hi) / 2;
  const fair = (p: number) =>
    (Math.sqrt(z * z + 4 * (1 - z) * ((p * p) / booksum)) - z) / (2 * (1 - z));
  return [fair(pOver), fair(pUnder)];
}

export type DevigMethod = "multiplicative" | "power" | "shin";

// Returns { fairOver, fairUnder, hold }. The two fair probs sum to 1; hold is
// the book's overround (raw implied-prob sum - 1).
export function devigTwoWay(
  overPrice: number,
  underPrice: number,
  method: DevigMethod = "multiplicative",
): { fairOver: number; fairUnder: number; hold: number } {
  const pOver = americanToProb(overPrice);
  const pUnder = americanToProb(underPrice);
  const hold = pOver + pUnder - 1;
  let fo: number;
  let fu: number;
  if (method === "multiplicative") [fo, fu] = multiplicative(pOver, pUnder);
  else if (method === "power") [fo, fu] = power(pOver, pUnder);
  else if (method === "shin") [fo, fu] = shin(pOver, pUnder);
  else throw new Error(`unknown devig method: ${method}`);
  return { fairOver: fo, fairUnder: fu, hold };
}

// Per-$1 EV of taking the UNDER at `offeredUnderPrice`, given a reference no-vig
// fair under probability. Positive = +EV.
export function evUnder(
  fairUnderProb: number,
  offeredUnderPrice: number,
): number {
  const payout = americanToDecimal(offeredUnderPrice) - 1;
  return fairUnderProb * payout - (1 - fairUnderProb);
}

// --- is this quote a centred line at all? ---------------------------------
//
// Mirrors beatvegas/devig.py::is_centred_quote -- see there for the full note
// and the measurement it came from.
//
// A book's MAIN total prices both sides near -110. An off-centre rung of the
// alternate ladder moves the line several points and goes lopsided to
// compensate, and the PRICE is the honest tell because it needs no reference.
//
// 2026 week 2, 1,504 pre-kickoff 1H quotes: no centred quote was worse than
// -150; off-centre rungs ran to -375 with a -250 median. -160 clears the
// observed centred range and still rejects 172 of 203 rungs.
export const SKEW_REJECT_PRICE = -160;

/**
 * True when both sides are priced like a book's main number. A quote missing
 * either side is treated as centred: this rejects a positively-identified
 * pathology, never data it cannot assess.
 */
export function isCentredQuote(
  overPrice: number | null | undefined,
  underPrice: number | null | undefined,
): boolean {
  for (const price of [overPrice, underPrice]) {
    if (price == null) continue;
    if (price < SKEW_REJECT_PRICE) return false;
  }
  return true;
}

// --- is this HARD ROCK quote its main line? -------------------------------
//
// Mirrors beatvegas/devig.py::is_hr_rung -- see there for the measurement.
// Hard Rock prices a 2-point alternate at about -145/-150 and a 3-point one at
// exactly -160, which SKEW_REJECT_PRICE lets through, and the Odds API has
// served one alone as Hard Rock's 1H total (Texas @ Tennessee 2026-09-25: 30.5
// at -160 while the app said 27.5). Either check is enough: the price, or the
// distance from the other books' median in the same sweep.
export const HR_RUNG_PRICE = -140;
export const HR_RUNG_DISTANCE_PTS = 2.0;
export const HR_RUNG_MIN_BOOKS = 3;

/**
 * True when a Hard Rock quote is an alternate line, not its main number.
 * `otherLines`: the other books' main lines as of the same moment; the
 * distance check needs HR_RUNG_MIN_BOOKS of them, else only the price counts.
 */
export function isHrRung(
  overPrice: number | null | undefined,
  underPrice: number | null | undefined,
  line: number | null | undefined,
  otherLines: readonly (number | null | undefined)[],
): boolean {
  for (const price of [overPrice, underPrice]) {
    if (price != null && price <= HR_RUNG_PRICE) return true;
  }
  const others = otherLines.filter((x): x is number => x != null);
  if (line == null || others.length < HR_RUNG_MIN_BOOKS) return false;
  const sorted = [...others].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  const med =
    sorted.length % 2 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2;
  return Math.abs(line - med) >= HR_RUNG_DISTANCE_PTS;
}

File: beatvegas/devig.py
"""No-vig (devigged) fair-probability math for two-way totals.

Kept dependency-free and unit-tested, mirrored verbatim by web/lib/devig.ts.
Convention: we reason about the UNDER side, but the math is symmetric.

The book's two posted prices imply probabilities that sum to >1 (the overround
/ hold). Devigging removes that margin to recover the market's fair probability.
Three standard methods:
  - multiplicative: divide each implied prob by their sum (fine for near-
    symmetric -110/-110 lines).
  - power: raise each implied prob to a common exponent k so they sum to 1
    (handles favorite-longshot skew better than multiplicative).
  - shin: Shin (1992/1993) insider-trading model; solves for the proportion of
    informed money z, then backs out fair probs.

Devig outputs are MARKET signal — they must never feed the market-blind BV
regressor. Used only for CLV, line-check EV, staking, and backtest grading.
"""

from __future__ import annotations

from typing import Tuple

from .grading import american_to_decimal


def american_to_prob(price: int) -> float:
    """Implied (vig-included) probability from American odds.

    -110 -> 0.5238, +120 -> 0.4545.
    """
    return 1.0 / american_to_decimal(price)


def _multiplicative(p_over: float, p_under: float) -> Tuple[float, float]:
    total = p_over + p_under
    return p_over / total, p_under / total


def _power(p_over: float, p_under: float) -> Tuple[float, float]:
    """Find k such that p_over**k + p_under**k == 1, via bisection."""
    lo, hi = 0.0, 10.0
    for _ in range(60):
        k = (lo + hi) / 2.0
        s = p_over**k + p_under**k
        if s > 1.0:
            lo = k  # larger k shrinks the sum (probs < 1)
        else:
            hi = k
    k = (lo + hi) / 2.0
    return p_over**k, p_under**k


def _shin(p_over: float, p_under: float) -> Tuple[float, float]:
    """Shin's two-outcome devig. The recovered fair probs are
        fair_i = (sqrt(z^2 + 4(1-z) * o_i^2 / booksum) - z) / (2(1-z))
    where o_i are the raw implied probs and z is the insider proportion. The
    sum-to-one constraint reduces (for two outcomes) to
        sqrt(z^2 + 4(1-z)a) + sqrt(z^2 + 4(1-z)b) = 2,   a=o_over^2/booksum, etc.
    which we solve for the interior root z in (0, 0.5) by bisection. Falls back
    to multiplicative when the overround is non-positive."""
    booksum = p_over + p_under
    if booksum <= 1.0:
        return _multiplicative(p_over, p_under)
    a = p_over**2 / booksum
    b = p_under**2 / booksum

    def s(z: float) -> float:
        return (z**2 + 4.0 * (1.0 - z) * a) ** 0.5 + (z**2 + 4.0 * (1.0 - z) * b) ** 0.5

    # s(0) = 2*sqrt(booksum) > 2; s dips below 2 before returning to 2 at z=1.
    # The wanted root is the first crossing, bracketed by [0, 0.5].
    lo, hi = 0.0, 0.5
    for _ in range(60):
        z = (lo + hi) / 2.0
        if s(z) > 2.0:
            lo = z
        else:
            hi = z
    z = (lo + hi) / 2.0

    def fair(p: float) -> float:
        return (((z**2 + 4.0 * (1.0 - z) * p**2 / booksum) ** 0.5) - z) / (2.0 * (1.0 - z))

    return fair(p_over), fair(p_under)


def devig_two_way(
    over_price: int, under_price: int, method: str = "multiplicative"
) -> Tuple[float, float, float]:
    """Return (fair_over, fair_under, hold) from a two-way total's prices.

    `hold` is the book's overround (raw implied-prob sum - 1). The two fair
    probabilities sum to 1.
    """
    p_over = american_to_prob(over_price)
    p_under = american_to_prob(under_price)
    hold = p_over + p_under - 1.0
    if method == "multiplicative":
        fo, fu = _multiplicative(p_over, p_under)
    elif method == "power":
        fo, fu = _power(p_over, p_under)
    elif method == "shin":
        fo, fu = _shin(p_over, p_under)
    else:
        raise ValueError(f"unknown devig method: {method}")
    return fo, fu, hold


def ev_under(fair_under_prob: float, offered_under_price: int) -> float:
    """Per-$1 EV of taking the UNDER at `offered_under_price`, given a reference
    no-vig fair under probability. Positive = +EV.

    EV = p*(decimal-1) - (1-p).
    """
    payout = american_to_decimal(offered_under_price) - 1.0
    return fair_under_prob * payout - (1.0 - fair_under_prob)
